File: Python/Easy/20-valid-parentheses/valid-parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        bracket_stack = []
        bracket_map = {")": "(", "}": "{", "]": "["}

        for char in s:
            if char in bracket_map:
                if len(bracket_stack) == 0 or bracket_stack.pop() !=  bracket_map[char]:
                    return False
                
            else:
                bracket_stack.append(char)
        
        return not bracket_stack

File: CSharp/Easy/20-valid-parentheses/valid-parentheses.cs
public class Solution {
    public bool IsValid(string s) {
        Stack<char> bracketStack = new Stack<char>();

        Dictionary<char, char> bracketMap = new Dictionary<char, char>{
            { ')', '(' },
            { '}', '{' },
            { ']', '[' }
        };

        foreach (char currentChar in s){
            if (bracketMap.ContainsKey(currentChar)){
                if (bracketStack.Count == 0 || bracketStack.Pop() != bracketMap[currentChar]){
                    return false;
                }
            }
            else{
                bracketStack.Push(currentChar);
            }
        }
        return bracketStack.Count == 0;
    }
}